### DISCUSSION TITLE
TTS Output

### DISCUSSION THREAD

**Discussion by nanderer** [Taewan-P/gpt_mobile#142] [Ideas]
Hi,

i can input stuff using my voice do to using google keyboard, but it would be really neet if there would be an output thingy that just hands the text generated to the systems default tts engine.

So but next to that copy button on each reply one speak button that dose two things, do some local shenanigans to determine the language the text is written in and then hand it over to the tts engine, afterwards convert the button to an pause button. Also do some notification play pause stuff and make sure the headset button is also a play pause thingy.

Also add an option to speak any reply and activate speach to text input afterwards to provide hands free  voice conversation with it if possible.

dont forget some settings where i can change pitch and speed like type and speek dose it fine https://f-droid.org/de/packages/com.googamaphone.typeandspeak/

thanks a lot  in advance  and keep up the good work

**Comment by nanderer** [comment]
its probably better to dont use this but its only an example on what i use to determine languages, build by mistral ai or gpt i dont remeber anymore, maybe there is some android native way to do so using the build in spell checker so you dont have to ship big packages but you could also make this optional to install after install the app i mean if possible...
[langpipe.zip](https://github.com/user-attachments/files/18217439/langpipe.zip)

(i use that to get subtitles read out and that only if they are in english, otherwise it goes crazy when playing some anime with karaoke)

**Comment by appatalks** [comment]
I am also play-grounding with native Android TTS functionality. 

Something like this appears to be working well for me:

<details>
  <summary>ChatBubble.kt</summary>

```js
    var tts: TextToSpeech? by remember { mutableStateOf(null) }
    val context = LocalContext.current
    var isTtsInitialized by remember { mutableStateOf(false) }
    LaunchedEffect(Unit) {
        tts = TextToSpeech(context) { status ->
            if (status == TextToSpeech.SUCCESS) {
                isTtsInitialized = true
                // setLanguageAndVoice(tts)
            } else {
                // Handle TTS Initialization failure
            }
        }
    }

...

    DisposableEffect(Unit){
        onDispose {
            tts?.stop()
            tts?.shutdown()
        }
    }
}
```
```js
...
                        RetryChip(onRetryClick)
                    }
                    if (isTtsInitialized) {
                        Text(
                            text = "\t\uD83D\uDD0A",
                            modifier = Modifier.clickable {
                                convertTextToSpeech(text, tts)
                            },
                            style = TextStyle(fontSize = MaterialTheme.typography.bodyLarge.fontSize * 1.65)
                        )
                    }
                }
            }
        }
    }
    DisposableEffect(Unit){
        onDispose {
            tts?.stop()
            tts?.shutdown()
        }
    }
}
```

```js
private fun setLanguageAndVoice(tts: TextToSpeech?) {
    // Use the system's default locale
    val desiredLocale = Locale.getDefault()
    tts?.setLanguage(desiredLocale)
    // Do not explicitly set the voice, allowing the system to use its default voice
    // tts.setVoice(selectedVoice); //Remove this line of code
}
```
```js
private fun convertTextToSpeech(text: String, tts: TextToSpeech?) {
    if (tts != null) {
        if (tts.isSpeaking) {
            tts.stop() // **Stop TTS if it's currently speaking**
        } else {
            tts.speak(text, TextToSpeech.QUEUE_FLUSH, null, null) // Start TTS if not speaking
        }
    }
```

</details>
